dialogue-target/src/main/java/com/palantir/dialogue/DefaultCallingThreadExecutor.java

```diff
@@ -0,0 +1,129 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.Preconditions;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RunnableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class DefaultCallingThreadExecutor implements CallingThreadExecutor {
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultCallingThreadExecutor.class);
+    private final long threadId = Thread.currentThread().getId();
+    private final Queue queue = new Queue();
+
+    /** Notification when main future completes. */
+    private final Runnable notifier = queue::poison;
+
+    @Override
+    public synchronized Future<?> submit(Runnable task) {
+        // if (Thread.currentThread().getId() == threadId) {
+        //     queue.checkNotPoisoned();
+        //     RunnableFuture<?> future = new FutureTask<>(task, null);
+        //     future.run();
+        //     return future;
+        // }
+
+        return queue.submit(task);
+    }
+
+    @Override
+    public void executeQueue(ListenableFuture<?> await) {
+        Preconditions.checkState(Thread.currentThread().getId() == threadId, "Executing queue on different thread");
+        Futures.addCallback(
+                await,
+                new FutureCallback<Object>() {
+                    @Override
+                    @SuppressWarnings("FutureReturnValueIgnored")
+                    public void onSuccess(Object _result) {
+                        queue.submit(notifier);
+                    }
+
+                    @Override
+                    @SuppressWarnings("FutureReturnValueIgnored")
+                    public void onFailure(Throwable _throwable) {
+                        queue.submit(notifier);
+                    }
+                },
+                DialogueFutures.safeDirectExecutor());
+
+        RunnableFuture<?> toRun;
+        while ((toRun = queue.getWork()) != null) {
+            toRun.run();
+        }
+    }
+
+    private static final class Queue {
+        private boolean poisoned = false;
+        private final BlockingDeque<RunnableFuture<?>> queue = new LinkedBlockingDeque<>();
```

Comment:
I think we can use a BlockingQueue rather than BlockingDequeue based on our access pattern.

Unsure how LinkedBlockingDequeue and LinkedBlockingQueue performance differs (no change requested, just thinking. Also wonder about LinkedTransferQueue, but that one spins before parking and eats some CPU)

In terms of state tracking and concurrency, I think we can remove synchronization entirely if we replace the `poisoned` access with a volatile boolean for `submit`, but add a sentinel value to the queue to signal `getWork` callers so they never read the `poisoned` field.

Reply:
Switched to BlockingQueue, I didn't even think twice about that. There's also ArrayListBlockingQueue: frankly that is probably the best for us, there shouldn't be more than 2 elements at any given time here right? Otherwise it's a bug. We can discuss

Reply:
re: removing synchronization: consider this ordering:


There's a blocking task in the queue to run the request.

Thread 1:
* calls #submit, checks volatile boolean, sleeps.

Thread 2:
# calls #executeQueue with the future.
# runs the task that is already in the queue, that completes the future, adds the notifier to the queue (or sentinel task).
# runs the notifier task, that sets the volatile boolean.
# there's no more tasks, the future is finished, so it exits.

Thread 1:
# wakes up, adds the task.

We have a hanging future?

Arguably, that is not really possible with the way the rest of dialogue works, but I'd rather write this in a bulletproof way.

Reply:
Ultimately, there should be zero contention for this lock; is it so bad to write it this way?

Reply:
Ya, the synchronization is probably fine, it should just be a cas to acquire the lock.